### ISSUE TITLE
mlp.Classifier._setup() fails to patch to sklearn's LabelBinarizer after pickle

### ISSUE THREAD

**Issue by yonil7** [aigamedev/scikit-neuralnetwork#161]
the following patch no longer work with new scikit-learn 0.17.0:
```python
def _setup(self):
    ...
    import sklearn.preprocessing.label as spl
    spl.type_of_target = lambda _: "multiclass"
```
This might be because scikit-learn type_of_target() function has moved to sklearn\utils\multiclass.py
because of this predict() and predict_proba() of a two class label classification problem returns wrong values.
e.g. predict_proba() returns array with a single columns (filled all with 1) instead of returning array with two columns.

**Comment by alexjc** [comment]
Hrm, I don't like the way this `sklearn` function is implemented` such context sensitive behavior is often fragile, should be more predictable (or at least a flag to make it work predictably). I wonder about just duplicating it...

**Comment by yonil7** [comment]
re: "I wonder about just duplicating it...". I think this is a good solution for the following reasons:
1. LabelBinarizer has pretty simple logic - meaning its relatively small and it doesnt seems that it will be changed a lot in future version of sklearn (e.g. bug fixes, enhancements,...)
2. sklean lib for example includes joblib third party lib inside (not depending on it) and this is even a bigger library.

re: "we could just change the monkey patching to use the new location too". I think the problem with that is if you have other place in your code that use any other sklearn functionality that depend on this function it will fail (maybe even without you knowing)

**Comment by alexjc** [comment]
@yonil7 I added [87453a0] to try to isolate this problem, with one assert and a new test, and can't reproduce this on the tip of sklearn.

What test code were you using?

**Comment by yonil7** [comment]
This happened to me only when I have an mlp.Classifier() instance pickled and then loaded and used in by another process. (I used joblib to fork different process and pass each one of them a classifier to fit in parallel)
I think I understand why this happens: Classifier._setup() method is called only when a new NeuralNetwork instance is created (its called from NeuralNetwork.__init__()).
when a pickled object is loaded its __init__ method is not called. so if you have a process that just loads a pickled classifier and try to fit it, the _setup() wont be called at all.

Maybe instead of duplicating it would be simpler to implement:
class MyLabelBinarizer(LabelBinarizer)
and override its 3 methods: fit(y), transform(y), inverse_transform(y) to transform the y somehow before or after calling the fit/transform methods of the parent class.
not sure.

**Comment by alexjc** [comment]
OK, that makes sense. Looking at a solution...
